src/peft/tuners/lora/loraga.py

```diff
@@ -0,0 +1,310 @@
+
+        # Save cache to disk if specified
+        if cache_file is not None:
+            cache: dict[str, Any] = {}
+            for name, module in get_target_modules(model, lora_config):
+                cache[f"{name}._peft_loraga_grad"] = module._peft_loraga_grad
+
+            os.makedirs(os.path.dirname(cache_file), exist_ok=True)
+            torch.save(cache, cache_file)
+
+
+def estimate_gradients(
+    model: nn.Module,
+    lora_config: LoraConfig,
+    train_step: Callable[[], None],
+):
+    """
+    Estimate gradients for LoRA-GA initialization.
+
+    This function enables gradient computation ONLY on target module weights and runs the train_step callback. This is
+    more memory-efficient than enabling gradients globally.
+    """
+    # Remember original training state
+    was_training = model.training
+    model.train()
+
+    # Get target modules list once for efficiency
+    target_module_list = list(get_target_modules(model, lora_config))
+
+    # Check if any supported layers were found
+    if not target_module_list:
+        raise ValueError(
+            "No supported layers found for LoRA-GA initialization. "
+            "LoRA-GA only supports nn.Linear and Conv1D layers. "
+            "Please ensure your model contains at least one of these layer types in target_modules."
+        )
+
+    # Initialize gradient storage for each target module
+    for name, module in target_module_list:
+        module._peft_loraga_grad = torch.zeros_like(module.weight.data, dtype=module.weight.dtype)
+        module._peft_loraga_grad_count = 0
+
+    # Memory-efficient gradient computation: disable gradients for all parameters first,
+    # then enable only for target module weights
+    original_requires_grad = {}
+    for name, param in model.named_parameters():
+        original_requires_grad[name] = param.requires_grad
+        param.requires_grad = False
+
+    # Enable gradients ONLY for target module weights
+    for name, module in target_module_list:
+        module.weight.requires_grad = True
+
+    # Register forward hooks to count gradient computations
+    # Each forward pass corresponds to one gradient computation
+    hooks = []
+
+    def create_forward_hook(module):
+        def forward_hook(m, input, output):
+            module._peft_loraga_grad_count += 1
+
+        return forward_hook
+
+    for name, module in target_module_list:
+        hook = module.register_forward_hook(create_forward_hook(module))
+        hooks.append(hook)
+
+    # Enable gradient computation and run train_step
+    with torch.enable_grad():
+        train_step()
+
+        # Accumulate gradients after all backward passes complete
+        for name, module in target_module_list:
+            if module.weight.grad is not None:
+                # Convert to same dtype as stored gradient
+                module._peft_loraga_grad += module.weight.grad.detach().to(module._peft_loraga_grad.dtype)
```

Comment:
I think this can be simplified:

```suggestion
    def backward_hook(module, grad_input, grad_output):
        module._peft_loraga_grad_count += 1

    for name, module in target_module_list:
        hook = module.register_full_backward_hook(backward_hook)
        hooks.append(hook)

    # Enable gradient computation and run train_step
    with torch.enable_grad():
        train_step()
```

1. No need to create a closure.
2. I think backward hook communicates the intent better and in theory, there could be multiple forward calls per backward.
3. The "Accumulate gradients after all backward passes complete" part doesn't really do anything useful or am I missing something?

Then below, do: `module._peft_loraga_grad = module.weight.grad.detach() / module._peft_loraga_grad_count` instead of `module._peft_loraga_grad /= module._peft_loraga_grad_count`. Let's also set the `.grad` to `None` afterwards just to be sure that memory can be freed.